backend/app/controllers/merge_request.rb

```diff
@@ -227,52 +279,137 @@ def merge_details(target, victim, selections, dry_run)
+      ind         = path_fix[1] 
+      mode        = path_fix[2]
+
+      subrec_id = values[subrec_name][ind]["id"]
+      subrec_index = find_subrec_index_in_victim(victim, subrec_name, subrec_id)
+
+      replacer = victim[subrec_name][subrec_index]
+
+      if mode == "replace"
+        target[subrec_name][ind] = process_subrecord_for_merge(target, replacer, subrec_name, mode, ind)
+      elsif mode == "append"
+        target[subrec_name].push(process_subrecord_for_merge(target, replacer, subrec_name, mode, ind))
+      end
+
+    end
+  end
+
+  # we don't know how the user reordered the subrecords on the merge form,
+  # so find the index with the right data given the ID of the right thing to replace/add by searching for it.
+  def find_subrec_index_in_victim(victim, subrec_name, subrec_id)
+    ind = nil
+    victim[subrec_name].each_with_index do |subrec, i|
+      if subrec["id"] == subrec_id
+        ind = i
+        break
+      end
+    end
+
+    return ind
+  end
+
+  # before we can merge a subrecord, we need to update the IDs, tweak things to prevent validation issues, etc
+  def process_subrecord_for_merge(target, subrecord, jsonmodel_type, mode, ind)
+    target_id = target['id']
+
+    if jsonmodel_type == 'names'
+      # an agent name can only have one authorized or display name.
+      # make sure the name being merged in doesn't conflict with this
+
+      # if appending, always disable fields that validate across a set. If replacing, always keep values from target 
+      if mode == "append"
+        subrecord['authorized']      = false
+        subrecord['is_display_name'] = false
+      elsif mode == "replace"
+        subrecord['authorized']      = target['names'][ind]['authorized']
+        subrecord['is_display_name'] = target['names'][ind]['is_display_name']
+      end
+
+    elsif jsonmodel_type == 'agent_record_identifiers'
+      # same with agent_record_identifiers being marked as primary, we can only have one
+
+      if mode == "append"
+        subrecord['primary_identifier'] = false
+
+      elsif mode == "replace"
+        subrecord['primary_identifier'] = target['agent_record_identifiers'][ind]['primary_identifier']
+      end
+    end
+
+    set_agent_id(target_id, subrecord)
+
+    return subrecord
+  end
+
+  # don't try to replace these values from victim to target when merging, ever!
+  def skippable_record_key?(k)
+    k == "agent_person_id" ||
+    k == "agent_family_id" ||
+    k == "agent_corporate_entity_id" ||
+    k == "agent_software_id" ||
+    k == "created_by" ||
+    k == "last_modified_by" ||
+    k == "create_time" ||
+    k == "system_mtime" ||
+    k == "user_mtime" ||
+    k == "lock_version" ||
+    k == "primary_identifier" || # only one primary identifier allowed in set 
+    k == "jsonmodel_type"
+  end
+
+  # needs to be passed into resolve_references.
+  def resolve_list
```

Comment:
Perhaps make this a constant array instead of having it be "hidden" in a method?